File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/typenodes/SimpleTypeNode.java
package org.vanautrui.languages.compiler.parsing.astnodes.typenodes;

import org.vanautrui.languages.compiler.lexing.tokens.TypeIdentifierToken;
import org.vanautrui.languages.compiler.lexing.utils.IToken;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;

public class SimpleTypeNode implements ITypeNode {

  public final String typeName;

  public SimpleTypeNode(TokenList tokens)throws Exception{
    IToken token = tokens.get(0);
    if (token instanceof TypeIdentifierToken) {
      this.typeName = ((TypeIdentifierToken) token).getContents();
      tokens.consume(1);
    } else {
      throw new Exception("could not read simple type identifier");
    }
  }

  @Override
  public String toSourceCode() {
    return this.typeName;
  }

  @Override
  public String getTypeName() {
    return this.typeName;
  }
}

File: dragon-parser/src/main/compiler/parsing/astnodes/typenodes/basic_and_wrapped/SimpleTypeNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "SimpleTypeNode.hpp"

public final class SimpleTypeNode implements IBasicAndWrappedTypeNode {

	public final String typeName;

	public SimpleTypeNode(final TokenList tokens) throws Exception {
		if (tokens.size() == 0) {
			throw new Exception();
		}

		IToken token = tokens.get(0);
		if (token instanceof TypeIdentifierToken) {
			this.typeName =
					token.getContents();

			tokens.consume(1);
		} else if (token instanceof AnyTypeToken) {
			tokens.consume(1);
			this.typeName = "#";
		} else {
			throw new Exception("could not read simple type identifier");
		}
	}

	public SimpleTypeNode(String typeName) {
		this.typeName = typeName;
	}


}
